enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,291 @@
+
+### Risks and Mitigations
+
+- Not all operators may take advanatage of this feature. If an operator does not opt-in to this feature, OLM will fallback on existing behavior when checking if an operator can be upgraded.
+
+- Operator may opt-in to this feature but never update the endpoint as "upgradeable". In these situations, cluster admins must uninstall the operator and upgrade the operator manually.
+
+- If OLM reports metrics to Telemeter regarding whether or not an operator is upgradeable an unbounded metric is being reported. An unbounded metric may overwhelm Telemeter.
+
+## Design Details
+
+### Test Plan
+
+OLM's e2e testing suite will be upgraded to handle the usecase:
+
+- If a CSV defines a upgradeReadiness Service without a spec that does not exist, upgrades for that operator are blocked.
+- If a CSVless bundle defines a upgradeReadiness Service that does not exist, the bundle fails to build.
+- If a CSV defines an upgradeReainess Service with a spec, changes to the service are reverted by OLM.
+- If a CSV/CSVless bundle defines an upgradeReainess Service that returns a 0, upgrades to the operator are blocked. The operator then upgrades when the service is no longer blocked.
+
+Additionally, tests that use CSVs that omit the upgradeReainess Service must continue to adhear to existing behavior.
+
+
+## Implementation History
+
+Major milestones in the life cycle of a proposal should be tracked in `Implementation
+History`.
+
+## Alternatives
+
+The are viable alternatives to the solution proposed in this enhancement.
+
+### Subsbitute the UpgradeReadinessService with an UpdateProbe
+
+It is possible that we could take advantage of the standard defined by [Liveness, Readiness, and StartUp Probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/) by importing the [Prober](https://github.com/kubernetes/kubernetes/blob/release-1.18/pkg/kubelet/prober/prober.go#L48) code. There are cons and benefits to this approach defined below:
+
+**This seems to be the most viable alternative.**
+
+#### Pros
+
+- Follows a well defined probe format which supports [HTTP Gets, EXEC, and TCPSocket](https://github.com/kubernetes/api/blob/6f652b6ce59c386f4a431eb031d0339620aaff5e/core/v1/types.go#L2279-L2292) options.
+- Enables OLM to identify which POD is reporting a upgradeReadiness failure, as a service could route to one of any number of pods.
+
+#### Cons
+- Requires OLM create an import and maintain prober code which executes the probe.
+
+### Create an API so Operators may report Status
+
+OLM could introduce a new API that allows operators to report their upgradeReadiness
+
+#### Pros
+
+- Provides a single source of truth for upgradeReadiness.
+
+#### Cons
+
+- Increases the number of APIs that OLM introduces. OLM has already received negative feedback on the number of APIs one must work with.
+- Behaves differently from other standard Kubernetes probes.
```

Comment:
I'm not sure how this is a con. It's normal to report status on a custom resource, often through conditions, and this would be a great fit. We don't need to think of this necessarily as a "probe".

Reply:
I would argue that this is a con:

- Requiring Operator Authors to edit the status introduces a higher barrier to entry than mimicking the Readiness/Liveness/StartUp readiness probe approach, which is often learned by even non-operator authors.
- Creating a new API creates additional "OLM domain specific knowledge", a complaint that many users have already voiced.

Reply:
I think it's worth discussing this option a bit more. There are "degrees" to this that we have discussed:

- Having a standard "upgradeable" condition, that operators can write into their _own_ APIs, which OLM then understands. 
- Having a an API which _OLM_ writes based on other inputs (update probes, standard conditions, config in the operator that says "I'm not upgreadable if you see X value in one of my CRs")
- Having an API (the same one) that operator authors opt-in to writing, if they choose to integrate with OLM more.

Reply:
The latest version of the PR now discusses this approach in much greater detail.